### PR TITLE
Use `-in-element` to render at appropriate time

### PR DESCRIPTION
**Pull request by rondale-sc**
This PR attempts to address the fact that in ember-cli-head's initializer we call `component.appendTo(document.head)`.  This triggers a rendering cycle where Glimmer fully processes the `head-layout` component and then appends it to the document head **before** the application's main rendering process.  

---

We decided not to do this:

~If we don't want to do a major version bump we need to do the following:~

- ~move head-browser-initializer to the addon folder~
  - ~Use [ember-compatibility-helpers](https://github.com/pzuraq/ember-compatibility-helpers) to remove the `component.appendTo(document.head)` only when `GLIMMER_2` is found~
- ~Use the `treeForFastboot` hook to emit the old fastboot files when Ember version of application is < 2.10~
- ~Use an AST transform to modify the user's application.hbs to include the invocation of `{{head-layout}}` component.~
  - ~In an ideal world we'd be able to do this more naively in a broccoli-plugin where we just find the actual template file and add it there, but unsure how/if we could do that.~

---

## Details

Currently it appends before the `didTransition` hook from the router happens, which means it is doing this work in the wrong phase of execution.  I did some benchmarking with ember-macro-benchmark and noticed this change when using this PR.  The experiment is our PR here.

screen shot 2017-08-03 at 2 23 56 pm

Which shows that we are moving the work to the correct place.  We can observe this more clearly in the Chrome timeline.

screen_shot_2017-08-03_at_2_39_14_pm

You can see in the above that ember-cli-head triggers a render before the main rendering pass for our application's first frame.

With this PR it looks more like this:

screen_shot_2017-08-03_at_2_46_15_pm

That's the TL;DR (which is too long)

But this is still very much WIP.  After discussing this with @rwjblue we decided that this would necessitate a major version bump unless we did a few things to make it backward compat.

## Benefits

I'd like to start a discussion about the possible benefits here.  With my benchmarking tools I couldn't see a pronounced difference (it isn't 100% clear yet).   Though, I would imagine there is some duplicated effort in starting the rendering process from a cold-start twice.

Any thoughts or ideas would be very much appreciated.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -26,6 +26,7 @@
     "ember-cli-app-version": "^1.0.0",
     "ember-cli-dependency-checker": "^1.3.0",
     "ember-cli-eslint": "^3.0.0",
+    "ember-cli-fastboot": "^1.0.2",
```
Comment (kratiahuja) [inline package.json]: Why this is added as a devDependnecy?

Comment (rondale-sc) [inline package.json]: Primarily to ensure this approach worked in fastboot.  Means that we can run `ember s` and disable JS then inspect the `head` element.  I can remove if that is a problem

Comment (rwjblue) [inline package.json]: Definitely makes for easier local testing.  Not sure if there are any ramifications to the ember-cli-addon-tests?

Comment (kratiahuja) [inline package.json]: Not an issue but I think you could write the tests in `fastboot-tests`? If it is more against the dummy app, it's fine too.

**Comment by rondale-sc** [comment]
This should be ready for another review.  🍻

File: addon/templates/components/head-layout.hbs [hunk 2]
```diff
@@ -1 +1,3 @@
-<meta name="ember-cli-head-start">{{head-content}}<meta name="ember-cli-head-end">
+{{#-in-element headElement}}
```
Comment (rwjblue) [inline addon/templates/components/head-layout.hbs]: FYI - We have started taking steps to make this public API from the Glimmer VM perspective (see https://github.com/glimmerjs/glimmer-vm/pull/619). It is unclear what Ember version that PR will correspond to, but it is a good step forward for our use case (and shows a future where we can drop the private `-` prefix).

**Comment by rondale-sc** [comment]
@rwjblue Added 2.12 to the ember-try config.  I'm not sure where to begin with the blueprint idea.

**Comment by rwjblue** [comment]
RE: the blueprint I'm suggesting that we update [the blueprint](https://github.com/ronco/ember-cli-head/blob/master/blueprints/ember-cli-head/index.js) to use [Blueprint.prototype.insertIntoFile](https://ember-cli.com/api/classes/Blueprint.html#method_insertIntoFile) to insert `{{head-layout}}`.

**Comment by rondale-sc** [comment]
@rwjblue That's an excellent idea.  Will give that a try.

**Comment by rondale-sc** [comment]
I think this is ready to go now.

File: blueprints/ember-cli-head/index.js [hunk 3]
```diff
@@ -1,7 +1,27 @@
...
+    let fullPath = path.join(this.project.root, 'app/templates/applicati.hbs');
+
+    if (fs.existsSync(fullPath)) {
+      let contents = fs.readFileSync(fullPath, { encoding: 'utf-8' });
+      let toWrite = `{{head-layout}}\n${contents}`;
+
+      fs.writeFileSync(fullPath, toWrite, { encoding: 'utf-8' });
+    } else {
+      let str = `You must add {{head-layout}} component to your topmost UI.
+This is usually your application.hbs, but was not found on your system.
```
Comment (rwjblue) [inline blueprints/ember-cli-head/index.js]: Maybe use the full path (e.g. `app/templates/application.hbs`) in the message?

File: blueprints/ember-cli-head/index.js [hunk 4]
```diff
@@ -1,7 +1,27 @@
...
+let fs = require('fs');
+let path = require('path');
+let chalk = require('chalk');
+
 module.exports = {
   description: '',
-  normalizeEntityName: function() { }
+  normalizeEntityName: function() { },
+  afterInstall() {
+    let fullPath = path.join(this.project.root, 'app/templates/applicati.hbs');
```
Comment (rwjblue) [inline blueprints/ember-cli-head/index.js]: Typo here (should be `app/templates/application.hbs`)

File: blueprints/ember-cli-head/index.js (hunk 3 above)
Comment (rondale-sc) [reply blueprints/ember-cli-head/index.js]: Updated!  🍻

**Comment by rwjblue** [comment]
LGTM, thanks for working through this with us!

@kratiahuja - Would you like to do a final review before landing and releasing?

**Comment by kratiahuja** [comment]
Sure I'll take a look at it in a while

**Comment by rondale-sc** [comment]
@kratiahuja Thank you for reviewing.  I added a section to the README and blueprint to help people moving from earlier versions.   I wasn't sure what the version would be for this so I guessed and used 0.4.x for the README section title, but I'll update that once this is ready to go.

👍

**Comment by rondale-sc** [comment]
Is there anything else blocking this before we cut a release?

**Comment by rwjblue** [comment]
Thank you for all of your hard work here @rondale-sc!

**Comment by Turbo87** [comment]
@rondale-sc @rwjblue am I interpreting correctly that this PR dropped support for Ember 2.4 and 2.8? might be worth it to mention that explicitly in the README and point to the v0.3.x releases for those Ember versions

**Comment by rwjblue** [comment]
@turbo87 - Yep, you are correct. Happy to review and land a PR updating the README...

**Comment by Turbo87** [comment]
Hmm I just noticed this in the README:

> Take into account that version >= 0.3 of this addon require Ember 2.10+ and fastboot >=1.0.rc1 Please use 0.2.X if you don't fulfull both requirements.

Since 0.3.x was still tested against Ember 2.4 and 2.8 is that statement actually valid? I'm a little confused... 🤔

**Pull request merged**